README
Comed - Comet server capable of high load

DESCRIPTION

This is a Perl implementation of Comet. The server was proven to withstand high
loads with more then 60000 simultanious user channels in production environment.
Comed performins authentication for clients against an external HTTP resource
based on client's session stored in cookies.


USAGE

  Start:
    $0 -c <config_file> [-d [-p <pid_file>] [-u <UID>] [-g <GID>] ]

  Stop:
    $0 -S -p <pid_file>




INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

DEPENDENCIES

Perl modules:

AnyEvent 5.21
AnyEvent::HTTP 1.45
EV 3.9
Try::Tiny 0.06
JSON::XS 2.26
Twiggy 0.1006
Config::General 2.48
Log::Dispatch 2.26

Libraries:

libev 3.9


COPYRIGHT AND LICENCE

Copyright (C) 2010, Alexey Skorikov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Changes history for Comed

1.88   Dec 16 14:47:17 2010
       Added ability to publish messages with a delay;
       Module bundling.

1.62   Dec 10 18:01:32 2010
       Statistics view.

1.62   Dec  8 19:31:55 2010
       Comed sender tool;
       Test coverage.

0.98   Nov 26 12:29:04 2010
       JSON deserialization.

0.82   Nov 21 18:06:14 2010
       Logging support.

0.60   Oct 29 18:12:52 2010
       Port reuse feature for Inbox.

0.46   Oct 25 11:37:18 2010
       Enhanced daemon capabilities;
       Enhanced MQ preformance.

0.10   Oct  5 14:26:53 2010
       Initial release.


lib/Comed.pm
package Comed;

use strict;
use 5.008_001;

use Carp;
use EV;
use AnyEvent;
use Twiggy::Server;
use Config::General qw( ParseConfig );
use Log::Dispatch;
use Comed::Log;
use Comed::MessageQueue;
use Comed::Inbox;
use Comed::Web;

our $VERSION = 1.88;

our %CONFIG_SETTINGS = (
    -LowerCaseNames => 1,
    -UseApacheInclude => 1,
    -IncludeRelative => 1,
    -IncludeDirectories => 1,
    -IncludeGlob => 1,
    -MergeDuplicateBlocks => 0,
    -MergeDuplicateOptions => 0,
    -AutoTrue => 0,
    -CComments => 0,
    -UTF8 => 0,
);

sub new
{
    my ($class, $config) = @_;
    my $ref = ref $config;

    (defined $config && (!$ref || ('HASH' eq $ref)))
        or croak 'Valid config HASHREF or filename SCALAR required';

    my $self = bless {
        config => $config,
        logger => undef,
    }, ref $class || $class;

    $self->read_config($config) unless $ref;
    return $self->init ? $self : 0;
}

sub config { shift->{config} }

sub init
{
    my $self = shift;
    my $config = $self->{config};

    my $log_level = $config->{logger}{level} || 'critical';
    my $logger = Comed::Log::set_logger( Log::Dispatch->new ) or return 0;

    if (my $log_file = $config->{logger}{file})
    {
        require Log::Dispatch::File;
        $logger->add( Log::Dispatch::File->new
            ( name      => 'logfile',
              min_level => $log_level,
              filename  => $log_file,
              mode      => '>>',
              newline   => 1,
            )
        );
    }

    if (my $syslog_ident = $config->{logger}{syslog})
    {
        require Log::Dispatch::Syslog;
        $logger->add( Log::Dispatch::Syslog->new
            ( name      => 'syslog',
              min_level => $log_level,
              ident     => $syslog_ident,
            )
        );
    }

    if ($config->{logger}{screen})
    {
        require Log::Dispatch::Screen;
        $logger->add( Log::Dispatch::Screen->new
            ( name      => 'screen',
              min_level => $log_level,
              newline   => 1,
            )
        );
    }

    $self->{logger} = $logger;
    $self->{cv}     = AE::cv;

    $self->{mq}     = Comed::MessageQueue->new(%{$config->{messagequeue}})
        or croak 'Failed to initialize message queue';

    $self->{inbox}  = Comed::Inbox->new(%{$config->{inbox}}, mq => $self->{mq})
        or croak 'Failed to initialize inbox';

    $self->{webserver}
                    = Twiggy::Server->new(%{$config->{web}{server}})
        or croak 'Failed to initialize web server';

    $self->{webapp} = Comed::Web->new(%{$config->{web}{app}}, mq => $self->{mq})
        or croak 'Failed to initialize web application';

    $self->{webserver}->register_service($self->{webapp}->psgi_app);

    1;
}

sub read_config
{
    my ($self, $filename) = @_;

    my %new_config = ParseConfig(
        -ConfigFile => $filename,
        %CONFIG_SETTINGS
    ) or croak 'Failed to read config file ' . $filename . '. ' . $!;

    $self->{config} = \%new_config;
}

sub start
{
    my $self = shift;
    $self->{cv}->recv;
}

1;
__END__


=head1 NAME

Comed - Comet server capable of high load


=head1 DESCRIPTION

This is a Perl implementation of Comet. The server was proven to withstand high
loads with more then 60000 simultanious user channels in production environment.
Comed is capable of performing authentication for clients against an external
HTTP resource based on session stored in cookies.


=head1 VERSION

This document describes Comed version 1.88


=head1 SYNOPSIS

    use Comed;

    my $comed = Comed->new( {
        messagequeue => {
            rotation_interval => 60,
        },
        inbox => {
            host    => '0.0.0.0',
            port    => 7777,
        },
        web => {
            server => {
                host    => '0.0.0.0',
                port    => 8888,
                timeout => 5,
            },
            app => {
                timeout          => 55,
                sidcookie_name   => 'ssid',
                sidcookie_length => 120,
                auth_uri_prefix  => 'http://my.app.com/authorize/?session=',
                auth_timeout     => 3,
                auth_status_ok   => 200,
            },
        },
    } );

    # Or

    my $comed = Comed->new('./comed.conf');

    # Then

    $comed->start;


lib/Comed/Log.pm
package Comed::Log;

use strict;
use POSIX qw(strftime);


our $LOGGER;
our %IMPORT_CALLED;


sub import
{
    my $class = shift;

    no strict qw(refs);

    my $caller_pkg = caller();

    return 1 if $IMPORT_CALLED{$caller_pkg}++;

    for(qw(DEBUG INFO NOTICE WARNING ERROR CRITICAL ALERT EMERGENCY)) {
        my $sub_name = $_;
        my $level    = lc($_);
        *{"$caller_pkg\::$sub_name"} = sub {
            length($_[0]) or return;
            my $message = strftime("[%d %b %Y %H:%M:%S]", localtime) .
                          " [$level] " . $_[0];
            $LOGGER ?
                $LOGGER->log(
                    level   => $level,
                    message => $message,
                ) :
                print STDOUT $message;
        };
    }
}

sub set_logger
{
    my $logger = shift;
    return $LOGGER = $logger if ((ref $logger) =~ /^Log::Dispatch/);
    return 0;
}

1;
__END__

lib/Comed/Web.pm
package Comed::Web;

use strict;

use AnyEvent::HTTP;
use JSON::XS;
use Try::Tiny;
use Comed::Log;
use Comed::MessageQueue;
use Data::Dumper;


our $VERSION = 1.88;


sub new
{
    my ($class, @args) = @_;

    my $self = bless {
        timeout              => 60,
        sidcookie_name       => 'session',
        sidcookie_length     => 128,
        auth_uri_prefix      => undef,
        auth_timeout         => 3,
        auth_status_ok       => 200,
        json_deserialize     => 0,
        num_connections      => 0,
        num_channel_requests => 0,
        num_auth_failures    => 0,
        num_errs_internal    => 0,
        @args,
    }, ref $class || $class;

    my $sidcookie_name   = $self->{sidcookie_name};
    my $sidcookie_length = $self->{sidcookie_length};
    $self->{sidcookie_regexp} = qr/$sidcookie_name=([^;]{1,$sidcookie_length})/;

    $self->{mq}   ||= Comed::MessageQueue->instance;
    $self->{json} ||= JSON::XS->new->allow_blessed
        if ($self->{json_deserialize});

    return $self;
}

sub num_connections { shift->{num_connections} }

sub throw
{
    my ($code, $msg) = @_;
    die [ $code, $msg ];
}

sub _catch
{
    my ($self, $error) = @_;

    if ('ARRAY' eq (ref $error)) {
        return [
            int($error->[0] || 500),
            [ 'Content-Type' => 'text/plain' ],
            [ $error->[1] || 'Unknown error' ]
        ];
    } else {
        ERROR $error;
        $self->{num_errs_internal} ++;
        return [
            500,
            [ 'Content-Type' => 'text/plain' ],
            [ "Internal Server Error" ]
        ];
    }
}

sub async_cb
{
    my $tx = shift;
    my $cb = shift;
    return sub {
        my @args = @_;
        try { $cb->(@args) }
        catch { $tx->[0]->croak($_) };
    };
}

sub reply_messages
{
    my ($self, $tx, $messages, $channel_id, $extra_headers) = @_;
    my $msg;

    if ($self->{json_deserialize})
    {
        my @deserialized = map { $self->{json}->decode($_) } @$messages;
        $msg = $self->{json}->encode(\@deserialized) if @deserialized;
    }
    else {
        $msg = '[' . join(',', @$messages) . ']' if @$messages;
    }

    $msg ||= '[[0]]';
    my $ct_postfix = $channel_id ? '; x-channel-id:' . $channel_id : '';
    my $headers = [
        'Content-Type'   => 'application/json' . $ct_postfix,
        'Content-Length' => length($msg),
    ];
    push @$headers, @$extra_headers if $extra_headers;

    $tx->[0]->send( [ 200, $headers ] );
    if (my $w = $tx->[1])
    {
        $w->write($msg);
        $w->close;
    }
}

sub reply_stats
{
    my ($self, $tx) = @_;
    my $now = time();
    my @uptime = gmtime($now - $^T);
    my $uptime = sprintf("%dd %dh %dm %ds", @uptime[7,2,1,0]);
    my $mq = $self->{mq};
    my $msg = <<END;
Comed Stats

--Comed---------------------------------
version          = r$VERSION
uptime           = $uptime

--Web-----------------------------------
connections      = ${\$self->{num_connections}}
channel_requests = ${\$self->{num_channel_requests}}
auth_failures    = ${\$self->{num_auth_failures}}
errors_internal  = ${\$self->{num_errs_internal}}

--MessageQueue--------------------------
last_channel     = ${\$mq->last_channel}
msgs_accepted    = ${\$mq->accepted}
msgs_delivered   = ${\$mq->requested}
msgs_rejected    = ${\$mq->rejected}
END

    $tx->[0]->send( [
        200, [
            'Content-Type'   => 'text/plain',
            'Content-Length' => length($msg),
        ], [ $msg ]
    ] );
}

sub reply_dump
{
    my ($self, $tx, $user_name) = @_;
    my $dump = Dumper($self->{mq}{b}[0]{$user_name});
    my $msg = <<END;
Comed MseeageQueue Dump, user '$user_name'

$dump
END

    $tx->[0]->send( [
        200, [
            'Content-Type'   => 'text/plain',
            'Content-Length' => length($msg),
        ], [ $msg ]
    ] );
}

sub process_request
{
    my ($self, $tx, $env) = @_;
    my $method = $env->{REQUEST_METHOD};
    my $path   = $env->{PATH_INFO};

    ('GET' eq $method) or throw(405, "Method not allowed\n");

    if ($self->{expose_stats})
    {
        return $self->reply_dump($tx, $1)
            if ($path =~ m|^/stats/dump/([^/]+)/?|);
        return $self->reply_stats($tx)
            if ($path =~ m|^/stats/?|);
    }

    my ($channel_id, $flags) = ($path =~ m|/poll/(\d{1,32})/(\d)|)
        or throw(404, "Not found\n");

    my ($session_cookie) = ($env->{HTTP_COOKIE} =~ $self->{sidcookie_regexp})
        or throw(403, "Session cookie needed\n");

    my @session_id_parts = split /:/, $session_cookie, 4;
    my $user_name = $session_id_parts[2];
    my $session_id = join '', @session_id_parts[0,1,3];

    my $mq = $self->{mq};
    my $cb = async_cb($tx, sub {
        $self->reply_messages($tx, $_[0], $channel_id)
    } );
    my $cv = AE::cv;
    $cv->cb(sub { $cb->($_[0]->recv) });

    if ($channel_id)
    {
        if ( my $messages =
                $mq->get_messages($channel_id, $user_name, $session_id, $cv) )
        {
            $cv->send($messages) if (@$messages);
        }
        else { $channel_id = 0 }
    }

    unless ($channel_id)
    {
        $self->{num_channel_requests} ++;

        if ($self->{auth_uri_prefix})
        {
            my $url = $self->{auth_uri_prefix} . $session_cookie;
            http_request(
                GET       => $url,
                timeout   => $self->{auth_timeout},
                on_header => async_cb($tx,
                sub {
                    my ($status) = ($_[1]{Status} =~ m/^(\d{3})/);
                    if ($self->{auth_status_ok} == $status)
                    {
                        $channel_id =
                            $mq->register_channel($user_name, $session_id, $cv)
                                or throw(500, "Channels' limit exceeded\n");
                    }
                    elsif (500 <= $status) {
                        ERROR sprintf
                            'Authorization request has failed: %s. (%s)',
                            $_[1]{Reason} || '', $url;
                        throw(403, "Forbidden\n");
                    }
                    else {
                        $self->{num_auth_failures} ++;
                        throw(403, "Invalid session\n");
                    }
                } ),
            );
        }
        else
        {
            $channel_id = $mq->register_channel($user_name, $session_id, $cv)
                or throw(500, "Channels' limit exceeded\n");
        }
    }

    my $t; $t = AE::timer $self->{timeout}, 0, sub {
        undef $t;
        $cv->send();
    };
}

sub run
{
    my ($self, $env) = @_;
    my $cv = AE::cv;
    my $tx = [ $cv ];

    return sub {
        my $start_response = shift;
        $tx->[0]->cb(sub {
            my $cv = shift;
            try {
                my $res = $cv->recv;
                my $w = $start_response->($res);
                if (!$res->[2] && $w) {
                    $tx->[1] = $w;
                }
            } catch {
                $start_response->($self->_catch($_));
            };
            $self->{num_connections} --;
        });

        $self->{num_connections} ++;

        try {
            $self->process_request($tx, $env);
        } catch {
            $cv->croak($_);
        };
    };
}

sub compile_psgi_app
{
    my $self = shift;
    return sub { $self->run(shift); }
}

sub psgi_app
{
    my $self = shift;
    return $self->{psgi_app} ||= $self->compile_psgi_app;
}

1;
__END__

lib/Comed/Inbox.pm
package Comed::Inbox;

use strict;
use AnyEvent;
use IO::Socket::INET;
use Comed::MessageQueue;


sub new
{
    my ($class, @args) = @_;

    my $self = bless {
        host       => '0.0.0.0',
        port       => 8686,
        reuse_port => 0,
        chunk_size => 1024,
        pattern    => undef,
        @args,
    }, ref $class || $class;

    $self->{socket} = new IO::Socket::INET(
        Proto     => 'udp',
        LocalHost => $self->{host},
        LocalPort => $self->{port},
        ReusePort => $self->{reuse_port},
        Broadcast => 1,
    ) or die $!;
    $self->{io}     = AE::io $self->{socket}, 0, sub { $self->read_message };
    $self->{mq}   ||= Comed::MessageQueue->instance;

    if (defined (my $pattern = $self->{pattern})) {
        $self->{pattern} = qr/^$pattern/ unless ref $pattern eq 'RegExp';
    }

    return $self;
}

sub read_message
{
    my $self = shift;
    recv $self->{socket}, my $raw, $self->{chunk_size}, 0;
    my ($user_name, $message) = split /,/, $raw, 2;
    my $pattern = $self->{pattern};

    $self->{mq}->accept_messages($user_name, $message)
        if (!$pattern || ($user_name =~ $pattern));
}

1;
__END__

lib/Comed/Daemon.pm
package Comed::Daemon;

use strict;
use warnings;

use Carp;
use Fcntl qw(:DEFAULT :flock);
use POSIX qw(setsid setuid setpgid);
use Exporter;
use base qw(Exporter);


our @EXPORT_OK = qw(start_daemon stop_daemon);
our %EXPORT_TAGS = (
    all => \@EXPORT_OK,
);


sub start_daemon
{
    my ($pid_file, $uid, $gid, $process_name) = @_;
    local $| = 1;

    $process_name ||= 'comed';

    print "Starting $process_name... ";

    if (my $old_pid = get_pid($pid_file)) {
        kill(0, $old_pid) and
            croak "The daemon is already running with PID $old_pid. " .
                  "Please, remove PID file $pid_file and try again.";
    }

    sysopen(PIDFILE, $pid_file, O_RDWR|O_CREAT) or
        croak "Can't open PID file $pid_file: $!";
    flock(PIDFILE, LOCK_EX | LOCK_NB) or
        croak "PID file $pid_file is already locked";
    sysseek PIDFILE, 0, 0 and truncate PIDFILE, 0 or
        croak "PID file $pid_file is not writable: $!";

    if ($uid || $gid)
    {
        chown(($uid || -1), ($gid || -1), $pid_file) or do
        {
            unlink $pid_file;
            croak "Can't change owner of PID file $pid_file: $!";
        };
    }

    my $attempt = 0;
    my $pid;

    while (not defined ($pid = fork())) {
        croak "Too many failed fork attempts: $!\n"
            if ++$attempt > 7;
        warn "Fork failed: $!\n";
        sleep 2;
    }

    if ($pid)
    {
        syswrite PIDFILE, "$pid\n", length("$pid\n") and close(PIDFILE)
            or croak "Can't write PID file $pid_file: $!";

        print "Done (PID=$pid)\n";
        exit 0;
    }

    umask 0;
    POSIX::setsid()                 or croak "Can't start a new session: $!";
    $gid and POSIX::setgid($gid)    || croak "Can't set GID: $!";
    $uid and POSIX::setuid($uid)    || croak "Can't set UID: $!";
    open STDIN,  q{<}, '/dev/null'  or croak "Can't read /dev/null: $!";
    open STDOUT, q{+>&STDIN}        or croak "Can't write to STDIN: $!";
    open STDERR, q{+>&STDIN}        or croak "Can't write to STDIN: $!";
    $0 = $process_name;

    $SIG{TERM} = $SIG{KILL} = sub
    {
        sysopen(PIDFILE, $pid_file, O_TRUNC);
        close(PIDFILE);
        unlink $pid_file;
        exit 0;
    };

    1;
}

sub stop_daemon
{
    my $pid_file = shift;
    local $| = 1;

    print "Stopping daemon... ";

    my $pid = get_pid($pid_file) or die
        "The daemon is probably not running. Failed to read PID file $pid_file";

    if ( kill(0, $pid) )
    {
        kill(15, $pid) or kill(9, $pid)
            or croak "Failed to kill the process $pid";
    }
    else
    {
        print "Failed to locate the process with PID $pid. " .
              "Probably it is not running. Cleaning PID file... ";
    }

    unlink $pid_file or croak "Failed to remove PID file $pid_file";
    print "Done\n";

    1;
}

sub get_pid
{
    my ($pid_file) = @_;

    (-r $pid_file) or return 0;
    open(PIDFILE, '<', $pid_file) or return 0;
    my $pid = <PIDFILE>;
    close(PIDFILE);
    return int($pid);
}

1;
__END__

lib/Comed/MessageQueue.pm
package Comed::MessageQueue;

use strict;
use AnyEvent;
use Math::BigInt;
use Try::Tiny;


my $MQ;


sub instance
{
    my $class = shift;
    $MQ = $class->new(@_) if @_;
    $MQ ||= $class->new;
}

sub new
{
    my ($class, @args) = @_;

    my $self = bless {
        rotation_interval   => 10, # seconds, fractional value
        buckets_number      => 2,  # >= 1
        max_store_msgs      => 64,
        max_user_chans      => 0,
        publish_async_delay => 0,
        channel_cnt         => Math::BigInt->new('0'),
        accepted_cnt        => Math::BigInt->new('0'),
        requested_cnt       => Math::BigInt->new('0'),
        rejected_cnt        => Math::BigInt->new('0'),
        b                   => [],
        @args,
    }, ref $class || $class;

    $self->rotate_buckets;
    $self->start_timer;

    return $self;
}

sub accepted           { shift->{accepted_cnt}->bstr  }
sub requested          { shift->{requested_cnt}->bstr }
sub rejected           { shift->{rejected_cnt}->bstr  }
sub last_channel       { shift->{channel_cnt}->bstr   }
sub reset_last_channel { shift->{channel_cnt}->bzero  }

sub rotate_buckets
{
    my $self = shift;
    my $buckets = $self->{b};

    unshift @$buckets, {};
    $#$buckets = $self->{buckets_number} - 1;

    return 1;
}

sub get_messages
{
    my ($self, $channel_id, $user_name, $cookie, $cv) = @_;
    my ($user, $bucket_id) = $self->_lookup_user($user_name);
    $user or return 0;
    my $channel = $user->{$channel_id} or return 0;

    # validation
    ( $cookie eq $channel->[1] ) or return 0;

    # retrieve and clear
    my $messages = delete $channel->[0];
    $channel->[0] = [];
    $channel->[2] = $cv if $cv;

    # refresh user
    $self->{b}[0]{$user_name} = $user if ($bucket_id != 0);

    my @messages_list = map { $$_ } @$messages;
    $self->{requested_cnt}->badd(scalar @messages_list);

    return \@messages_list;
}

sub accept_messages
{
    my ($self, $user_name, @messages) = @_;
    my $num_messages  = @messages or return 0;
    my ($user)        = $self->_lookup_user($user_name);

    if (!$user)
    {
        $self->{rejected_cnt}->badd($num_messages);
        return 0;
    }

    my $messages      = \@messages;
    my @channels      = values %$user;
    my $req_chans_cnt = 0;
    my $async_delay   = $self->{publish_async_delay};

    for my $id (0..$#channels)
    {
        if ($req_chans_cnt && $async_delay)
        {
            my $t; $t = AE::timer $async_delay, 0, sub
            {
                undef $t;
                my $cnt = 0;
    
                for my $id_t ($id..$#channels)
                {
                    $self->_publish_messages($channels[$id_t], $messages) &&
                        $cnt++;
                }
    
                $self->{requested_cnt}->badd($cnt * $num_messages) if $cnt;
            };
    
            last;
        }
    
        $self->_publish_messages($channels[$id], $messages) &&
            $req_chans_cnt++;
    }

    $self->{accepted_cnt}->badd($num_messages);
    $self->{requested_cnt}->badd($req_chans_cnt * $num_messages)
        if $req_chans_cnt;

    return 1;
}

sub register_channel
{
    my ($self, $user_name, $cookie, $cv) = @_;

    my $bucket = $self->{b}[0];
    my $user = $bucket->{$user_name} ||= {};
    my $num_channels = keys %$user;

    if (my $max_user_chans = $self->{max_user_chans})
    {
        ($num_channels >= $max_user_chans) && return 0;
    }

    my $channel_id = $self->{channel_cnt}->binc->bstr;
    $user->{$channel_id} = $cv ? [ [], $cookie, $cv ] : [ [], $cookie ];

    return $channel_id;
}

sub start_timer
{
    my $self = shift;

    $self->{timer} = AnyEvent->timer(
        after    => $self->{rotation_interval},
        interval => $self->{rotation_interval},
        cb       => sub { $self->rotate_buckets },
    );
}

sub stop_timer
{
    my $self = shift;
    undef $self->{timer};
}

sub _publish_messages
{
    my ($self, $channel, $messages) = @_;
    my $cv = $channel->[2];
    my $need_backup = 1;

    if ($cv && $cv->cb)
    {
        $need_backup = 0;
        try {
            $cv->send($messages);
        } catch {
            $channel->[2] = undef;
            $need_backup  = 1;
        }
    }

    if ($need_backup)
    {
        my $msgs_upper   = $self->{max_store_msgs} - 1;
        my $message_pool = $channel->[0];

        push @$message_pool, map { \$_ } @$messages;

        $#$message_pool = $msgs_upper
            if (($msgs_upper >= 0) && ($#$message_pool > $msgs_upper));

        return 0;
    }

    return 1;
}

sub _lookup_user
{
    my ($self, $user_name) = @_;
    my ($user, $bucket_id);
    my $buckets = $self->{b};

    for (0..$#$buckets)
    {
        $bucket_id = $_;
        $user = $buckets->[$bucket_id]{$user_name} and last;
    }

    return ($user, $bucket_id);
}

sub _buckets { shift->{b} }

1;
__END__

Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Comed',
    AUTHOR              => 'Alexey Skorikov <alexey@skorikov.name>',
    VERSION             => 1.88,
    ABSTRACT_FROM       => 'lib/Comed.pm',
    PL_FILES            => {},
    MIN_PERL_VERSION    => 5.8.8,
    PREREQ_PM => {
        'AnyEvent'            => 5.21,
        'AnyEvent::HTTP'      => 1.45,
        'EV'                  => 3.9,
        'Twiggy'              => 0.1006,
        'Try::Tiny'           => 0.06,
        'JSON::XS'            => 2.26,
        'Config::General'     => 2.48,
        'Log::Dispatch'       => 2.26,
        'Test::More'          => 0,
        'Test::Pod'           => 1.14,
        'File::Temp'          => 0,
        'version'             => 0,
    },
    PREREQ_FATAL        => 1,
    EXE_FILES           => [qw| script/comed script/comedsend |],
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Comed-*' },
);
